### DISCUSSION TITLE
登録したデータが表示されない

### DISCUSSION THREAD

**Discussion by kzkondo** [dentsusoken/iPLAss#1266] [Q&A]
この度ローコード開発の新たな候補として、
iPLAssを導入しようと考えており、試験的に環境を構築しました。
Dockerでコンテナも作成できましたが、
データを登録しても表示されません。
データベースの中身を確認してみるとレコードは登録されているのですが、
GEMモジュールに表示されません。
<img width="1680" alt="スクリーンショット 2023-08-20 13 25 39" src="https://github.com/ISID/iPLAss/assets/93172658/cb0d2318-e630-40df-a562-7c495be23515">
<img width="1680" alt="スクリーンショット 2023-08-20 13 25 54" src="https://github.com/ISID/iPLAss/assets/93172658/2fe735be-9309-43bf-ab85-848145affefc">

検索ボタンを押すと上記のようなエラーが出ます。

環境としては、
Dockerイメージ：iPLAss / iPLass 3.2.9
データベース：PostgreSQL 15.4.1
JDBC：JAVA8 42.6.0
で構築しています。

GEMモジュールの左メニューにある、
・権限情報
・基本情報
・お知らせ情報
の各項目においても新規登録しても、
同じ症状になっており、
登録されたレコードが表示されません。（こちらもデータが登録できていることは確認できます）

原因として何が悪いのか、どこを改善すれば良いのか、色々試してみましたが、
結果は同じのため分かりません。

素人なみの質問をしていることと存じておりますが、
何卒ご回答のほどよろしくお願いいたします。

**Comment by HiguchiKiyoshi** [comment]
@kzkondo 
iPLAssにご関心を寄せていただき、ありがとうございます。
こちらでも同様の事象が発生するか動作確認させていただきますので、お時間いただければと思います。
もし、エラー時のログなどありますようでしたら、ご提供いただけますと助かります。

**Comment by kzkondo** [comment]
@HiguchiKiyoshi さん、ご返信ありがとうございます。
最新のエラーをはいたHTMLを添付しますね。
よろしくお願いいたします。
[iplass.html 18-32-22-916.html.zip](https://github.com/ISID/iPLAss/files/12407137/iplass.html.18-32-22-916.html.zip)

**Comment by HiguchiKiyoshi** [comment]
@kzkondo
ログありがとうございます。
ログなどから推察するに、今回試験的に構築した環境はローカル端末上ではなく、クラウド環境上などに構築した形でしょうか。
例えば、iPLAssのコンテナの前段にロードバランサなど配置し、そこでTLSのオフロード処理など実施してる構成でしょうか？

もしそうであるならば、
そのオフロード処理とTomcatのConnector周りの設定（server.xmlに設定）が影響しているかもしれません。

※こちらの手元のローカル環境上でご提示いただいた構成（iPLAssコンテナ＋Postgresql）では事象は再現しませんでした。

**Reply by kzkondo** [reply]
@HiguchiKiyoshi さん
ログ解析ありがとうございます。
おっしゃっておられるのが、
リバースプロキシのことをおっしゃられておられるなら、
そのような処理を行っております。

ただ、ローカルへのアクセス、例えば、
http://192.168.0.2:8080/iplass
のようなアクセスの場合リバースプロキシを介さないので、
関係なくなると思うのですが、
このようなアクセスの場合も同じ現象が発生します。

そもそも僕の見解が間違っているかもしれませんので、
ご指摘お願いいたします。

もし見解が正しいのであれば何が問題として考えられるかご教示いただけると嬉しいです。

**Reply by HiguchiKiyoshi** [reply]
@kzkondo
はい、質問させていただいた内容は、kzkondoさんのご理解の通りです。

発生している事象が、
「画面は表示される、WebAPI呼び出しはエラーになる」
といった挙動から、
CORS（Cross-Origin Resource Sharing）による制限に引っかかっているのかもと疑いました。

iPLAss内のWebAPIの処理でCORSに基づき、クロスドメイン呼び出しかどうかのチェックを実装しています。
このとき、クロスドメイン通信となっているかどうかの比較にサーバサイドのロジック内（コンテナ内のTomcat内のロジック）から、
呼び出しされたURLのサーバ名、ポート、スキーム（http or https）を取得し確認しています。

この処理を正常に処理させるために、
外部から見えるサーバ名、ポート、スキームがサーバサイドのロジック内からも同様の値が取得できる必要があります。
リバプロやLBを挟む場合、この値がずれることがあるので、正しく設定してあげる必要があります。

が、なるほど、、、ローカルへの直アクセスでも動きませんか。。。
http://localhost:8080/iplass
でもだめでしょうか？

**Reply by kzkondo** [reply]
@HiguchiKiyoshi さん
ご提案ありがとうございます。
やはりそうですよね。
ローカルIPアドレスなら問題ないと思っておりましたが、
考え方は間違っていなかったようで良かったです。

ご提案いただいた、
http://localhost:8080/iplass
はNAS（Synology DSM7.2）のためブラウザをもっておらず、
上記でのアクセスする手段が考えつきません。

もし上記アクセスが解決策なのであれば試したいですが、
よい方法をご教示ください。

**Reply by HiguchiKiyoshi** [reply]
@kzkondo
理解しました。NAS（Synology DSM7.2）上で動かそうとしてたのですね。
上記アプライアンス製品詳しくないので軽くググった程度＆憶測で申し訳ありませんが、
IPアドレス指定であっても、おそらくリバプロ経由のアクセスになるのだと思います。

解決策としては、次のような手段が考えられます。

A案:
（単なる動作検証目的であれば）
NAS上ではなく、ローカル端末（Maｃなどで可）にDocker Desktopをインストールしていただき、ローカル端末上でiPLAssコンテナ（iplass/iplass-standalone）を稼働させ、localhost指定で直アクセスしていただく。

https://qiita.com/yuma_ishida_iPLAss/items/0c14b62457983b1c6ab0
https://iplass.org/docs/gettingstarted/docker/index.html
が参考になると思われます。

B案:
（サーバで稼働させたい場合）
TomcatのConnector設定をリバプロに合わせ適切に設定する。
/usr/local/tomcat/conf/server.xml内Connector要素を編集。
proxyName, proxyPort属性を外部公開されているサーバ名、ポートに合わせて設定。

参考URL
https://tomcat.apache.org/tomcat-9.0-doc/config/http.html
※すみません、ググっていただけたらもう少しわかりやすいサイトもあると思われます。
　公式サイト以外のリンクは控えさせていただきます。

コンテナ内の設定ファイルの修正は、/usr/local/tomcat/conf/をバインドマウント（ホスト側のファイルシステムにマウント）していただき、
直接書き換えていただく形が一番手っ取り早いと思います。
NAS（Synology DSM7.2）でも、「ボリューム設定」でマウントできそうです（ググった感じ）

**Reply by kzkondo** [reply]
@HiguchiKiyoshi さん
色々とお調べくださりありがとうございました。
Tomcatの設定ファイルを編集しようと、
/usr/local/tomcat/conf
をマウントするため、
コンテナを削除し新たにインストールし直してみました。

ところがマウントしようとして実行すると、
何度やってもiPLAssのコンテナが立ち上がってすぐに停止するため、
やむなくいつも通り設定し試しに、
IPアドレスでアクセスしてみました。
http://192.168.0.2:8080/iplass

すると前回は登録しているデータにアクセスできなかったものに、
アクセスできるようになりました。
これは今回上手くいったのではなく、元々うまくいっていた可能性は拭い去れません。
僕のチェックが甘かった可能性があります。

相変わらず、リバースプロキシを通じての症状は同じですが、
http://localhost:8080/iplass
を試してくださいと言っていただく前の状態まではうまく表示できていますので、
ここから先どこの部分に手を加えれば良いかご教示いただけますでしょうか？

**Reply by HiguchiKiyoshi** [reply]
@kzkondo
リバースプロキシ経由する場合はやはりTomcatの設定ファイルを正しく設定する必要があります。

エラーは推察で申し訳ないですが、ホスト側にTomcatの設定ファイルが存在していないためと思われます。
設定ファイルをコピーしてあげる必要がありますが、そちらは実施された形でしょうか？
手順の概略は以下になります。

①コンテナ内の conf ディレクトリをホストOSへコピー
②ホストOSへコピーした、server.xml を修正する
③ホストOSへコピーした conf ディレクトリをコンテナの conf ディレクトリへマウント

実際のコピーの仕方はNAS製品ですと可能かどうか不明ですが、一般的にはdockerコマンドを利用します。
※ググっていただくとわかりやすい説明があると思われます。

**Reply by kzkondo** [reply]
@HiguchiKiyoshi さん
ご教示くださり、ありがとうございます。

>設定ファイルをコピーしてあげる必要がありますが、そちらは実施された形でしょうか？

こちらを実施していないのでエラーになっているのですね。

一応、confディレクトリのコピーまでは完了し、
server.xmlも編集できるところまではきたのですが、
どこを編集したらいいのかが分からず、
proxyと記述のある部分も無いため教えてくださいますと助かります。

**Reply by HiguchiKiyoshi** [reply]
@kzkondo 
デフォルトのserver.xmlでは、proxyは利用しない形なので設定を追記していただく形になろうかと思います。

プロキシでマッピングしているポートが8080の場合、
```
<Connector port="8080" protocol="HTTP/1.1"
               connectionTimeout="20000"
               redirectPort="8443" />
```
と定義されているConnector要素に、proxyName、proxyPort、scheme、secureなどの属性を追加します。
例えば、次のような形（httpsで公開してる場合）にします。
```
<Connector port="8080" protocol="HTTP/1.1"
               connectionTimeout="20000"
               redirectPort="8443"
               proxyName="[外部公開される際のサーバ名]"
               proxyPort="443"
               scheme="https"
               secure="true"
               />
```
具体的な設定値については以下を参考にしてください。
https://tomcat.apache.org/tomcat-9.0-doc/config/http.html
※すみません、ググっていただけたらもう少しわかりやすいサイトもあると思われます。
　公式サイト以外のリンクは控えさせていただきます。

**Reply by kzkondo** [reply]
@HiguchiKiyoshi さん
丁寧に解説いただき、ありがとうございました。
無事、リバースプロキシを利用しアクセスすることができました。
これでテスト環境が整いましたので、
今後の開発に適しているかどうか、
あちこち触らせていただきます。
